src/libsyntax/feature_gate.rs

```diff
@@ -463,30 +573,32 @@ impl<'a, 'v> Visitor<'v> for PostExpansionVisitor<'a> {
     }
 
     fn visit_attribute(&mut self, attr: &ast::Attribute) {
-        if attr.check_name("staged_api") {
-            self.gate_feature("staged_api", attr.span,
-                              "staged_api is for use by rustc only");
-        } else if attr.check_name("plugin") {
-            self.gate_feature("plugin", attr.span,
-                              "compiler plugins are experimental \
-                               and possibly buggy");
-        }
-
-        if attr::contains_name(slice::ref_slice(attr), "lang") {
-            self.gate_feature("lang_items",
-                              attr.span,
-                              "language items are subject to change");
-        }
-
-        if attr.check_name("no_std") {
-            self.gate_feature("no_std", attr.span,
-                              "no_std is experimental");
-        }
-
-        if attr.check_name("unsafe_no_drop_flag") {
-            self.gate_feature("unsafe_no_drop_flag", attr.span,
-                              "unsafe_no_drop_flag has unstable semantics \
-                               and may be removed in the future");
+        match &*attr.name() {
+            "staged_api" => self.gate_feature("staged_api", attr.span,
+                                              "staged_api is for use by rustc only"),
+            "plugin" => self.gate_feature("plugin", attr.span,
+                                          "compiler plugins are experimental \
+                                           and possibly buggy"),
+            "no_std" => self.gate_feature("no_std", attr.span,
+                                          "no_std is experimental"),
+            "unsafe_no_drop_flag" => self.gate_feature("unsafe_no_drop_flag", attr.span,
+                                                       "unsafe_no_drop_flag has unstable \
+                                                        semantics and may be removed \
+                                                        in the future"),
+            "lang" => self.gate_feature("lang_items",
+                                        attr.span,
+                                        "language items are subject to change"),
+            name => {
+                // Custom attribute check
+                if KNOWN_ATTRIBUTES.iter().all(|&(n, _)| n != name) {
+                    self.gate_feature("custom_attribute", attr.span,
+                               format!("The attribute `{}` is currently \
+                                        unknown to the the compiler and \
+                                        may have meaning \
+                                        added to it in the future",
+                                        attr.name()).as_slice());
```

Comment:
The interaction between https://github.com/rust-lang/rust/pull/22336 and this is interesting. I think I'd like a way to include something in the table of custom attributes but still declare that it requires a (perhaps generic) feature-gate. This would be the case for things like `rustc_move_fragments` and `rustc_variance`. Or we could add them to the list above, but it seems a bit redundant. (It might also be nice to be able to just write something like `FeatureGate("tag")` and get a generic message like "The attribute `{}` has unstable semantics and requires the feature-gate `{}`".